src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -346,6 +346,16 @@ impl<'tcx> LoanPath<'tcx> {
     }
 
     fn to_type(&self) -> Ty<'tcx> { self.ty }
+
+    fn is_downcast(&self) -> bool {
```

Comment:
Nit: maybe call this `has_downcast`? Also, we should try this out with the MIR borrowck.